### PR TITLE
[github-actions] Generate tools cache for each PR

### PR DESCRIPTION
**Pull request by djaglowski**
Updates #9280

This PR changes the scope of caching in github actions. Basically, it localizes the cache to a per-PR level. The benefit of this is that when a cache becomes corrupted, it only affects a single PR.

There are some downsides and limitations to this change.
1. This does not solve the underlying cause of corrupted caches, which appears to be out of our control (or at least is not yet understood). 
2. It also does not provide the average PR author with an obvious solution to the problem. 
3. It adds a once-per-PR (ie. not once-per-CI-run) cost to generate and upload the cache.

In my opinion, these costs & limitations are preferable to the current state of things, where a corrupted cache blocks all new PRs on the entire repo until a new cache is forced. 

As a lesser benefit, this provides a self-service workaround for PR authors, in the sense that closing the affected PR and opening a new one should resolve the issue. This of course requires that the author is able to identify the issue, which is difficult unless familiar with the symptoms.

### CODE REVIEW

**Comment by djaglowski** [comment]
Note the PR number at the end of the cache, observed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/6575103423?check_suite_focus=true#step:13:5).

image

File: .github/workflows/prometheus-compliance-tests.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
         with:
           repository: prometheus/compliance
           path: compliance
-          ref: f0482884578bac67b053e3eaa1ca7f783d146557
```
Comment (bogdandrutu) [inline .github/workflows/prometheus-compliance-tests.yml:38]: Why this change?

File: pkg/stanza/go.sum [hunk 2]
```diff
@@ -288,8 +288,8 @@ go.opencensus.io v0.22.3/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.22.4/go.mod h1:yxeiOL68Rb0Xd1ddK5vPZ/oVn4vY4Ynel7k9FzqtOIw=
 go.opencensus.io v0.22.5/go.mod h1:5pWMHQbX5EPX2/62yrJeAkowc+lfs/XD7Uxpq3pI6kk=
 go.opencensus.io v0.23.0/go.mod h1:XItmlyltB5F7CS4xOC1DcqMoFqwtC6OG2xF7mCv7P7E=
-go.opentelemetry.io/collector v0.51.0 h1:r4cfvLYkmdTWU+RioL2QqxD/GLJUXRNifG0NCfRmxwk=
-go.opentelemetry.io/collector v0.51.0/go.mod h1:2pHwJxKDhSd4t42GkiCosU9G3szHoWUK2R0qnB6X67A=
```
Comment (bogdandrutu) [inline pkg/stanza/go.sum:292]: Why this change?

**Comment by TylerHelmuth** [comment]
> @TylerHelmuth, [here's an example](https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/6561219153?check_suite_focus=true#step:7:364) of what the failure looks like. Basically, jobs will fail when they try to use tools that should have been cached.

I see how identifying the issue as a cache corruption is kinda tricky :/

File: pkg/stanza/go.sum (hunk 2 above)
Comment (djaglowski) [reply pkg/stanza/go.sum:292]: Necessary to pass CI. It was missed in another PR due to the module being new.

I've removed the change and will resolve it elsewhere.

**Comment by djaglowski** [comment]
> @djaglowski this PR will not fix that problem that is linked. Since the upload failure can happen even with this PR.

Correct. I've noted this in the PR description. The proposal is to reduce the impact when the problem occurs.

**Comment by djaglowski** [comment]
image

One of the jobs on this PR is hanging due to what appears to be a different cache issue. I'm going to convert this back to a draft until I have a better understanding of this.
